uaclient/cli.py

```diff
@@ -1073,22 +1083,72 @@ def action_auto_attach(args, *, cfg):
 @assert_root
 @assert_lock_file("ua attach")
 def action_attach(args, *, cfg):
-    if not args.token:
+    if not args.token and not args.attach_config:
         raise exceptions.UserFacingError(
             ua_status.MESSAGE_ATTACH_REQUIRES_TOKEN
         )
-    try:
-        actions.attach_with_token(
-            cfg, token=args.token, allow_enable=args.auto_enable
+    if args.token and args.attach_config:
+        raise exceptions.UserFacingError(
+            ua_status.MESSAGE_ATTACH_TOKEN_ARG_XOR_CONFIG
         )
+
+    enable_services_override = None
+    if args.token:
+        token = args.token
+    else:
+        try:
+            attach_config = AttachActionsConfigFile.from_dict(
+                yaml.safe_load(args.attach_config)
+            )
+        except IncorrectTypeError as e:
+            raise exceptions.UserFacingError(
+                textwrap.fill(
+                    "Error while reading {}: {}".format(
+                        args.attach_config.name, e.msg
+                    ),
+                    width=PRINT_WRAP_WIDTH,
+                )
+            )
+        token = attach_config.token
+        if attach_config.enable_services is not None:
+            enable_services_override = attach_config.enable_services
+
+    allow_enable = args.auto_enable and enable_services_override is None
+
+    try:
+        actions.attach_with_token(cfg, token=token, allow_enable=allow_enable)
     except util.UrlError:
         print(ua_status.MESSAGE_ATTACH_FAILURE)
         return 1
     except exceptions.UserFacingError:
         return 1
     else:
+        ret = 0
+        if enable_services_override is not None:
+            for name in enable_services_override:
+                ent_cls = entitlements.entitlement_factory(name)
+                if ent_cls is None:
+                    logging.warning(
+                        ua_status.MESSAGE_INVALID_SERVICE_OP_FAILURE_TMPL.format(  # noqa: E501
+                            operation="enable", name=name, service_msg=""
+                        )
+                    )
+                    ret = 1
+                    continue
+                entitlement = ent_cls(
+                    cfg, assume_yes=True, allow_beta=True, called_name=name
+                )
+                ent_ret, reason = entitlement.enable()
```

Comment:
Should we adapt the logic we have on `action_enable` here somehow ?

Reply:
It looked a little different, but a small refactor might make it reusable, I'll look into it

Reply:
@lucasmoura I did a small refactor to share this code, let me know what you think!